core/src/worker/activities/activity_heartbeat_manager.rs

```diff
@@ -50,7 +53,11 @@ enum HeartbeatExecutorAction {
     /// Heartbeats are throttled for this task token, sleep until duration or wait to be cancelled
     Sleep(TaskToken, Duration, CancellationToken),
     /// Report heartbeat to the server
-    Report(TaskToken, Vec<common::Payload>),
+    Report {
+        task_token: TaskToken,
+        details: Vec<common::Payload>,
+        on_reported: Option<Arc<Notify>>,
```

Comment:
Ehh, I tried it and it's the same LOC and not obviously cleaner to me. You have to maintain another mapping like `tts_to_final_completes` because evict removes from `tt_to_state`, which is a little more action-at-a-distance-y than everything going in the messages.